### PR TITLE
Smt rw tac: Added the import to Smt.Reconstruction.Certifying

### PR DESCRIPTION
**Pull request by mhk119**

### CODE REVIEW

**Comment by tomaz1502** [comment]
This is looking good! Two comments about the last commit (please make a PR before pushing to main, so its easier to review):

- Its not necessary to do ```replaceMainGoal [mv]``` at the end of ```evalSMTRw```. The goal is already closed when you do ```mv'.refl```
- The variable ```m``` inside both for-loops in ```smtRw``` does not need to be mutable

**Comment by tomaz1502** [comment]
Also, in ```smtRw``` you could iterate through arrays instead of indices to avoid unsafe operations, like:
```lean
  for innerArr in arr do
    let m := innerArr.size
    if m > 1 then
      for j in [: m-1] do
        let r ← mv'.rewrite (← mv'.getType) (mkAppN assoc #[innerArr[m-j-2]!]) true
        mv' ← mv'.replaceTargetEq r.eNew r.eqProof
```

**Comment by mhk119** [comment]
> This is looking good! Two comments about the last commit (please make a PR before pushing to main, so its easier to review):
> 
> * It's not necessary to do `replaceMainGoal [mv]` at the end of `evalSMTRw`. The goal is already closed when you do `mv'.refl`
> * The variable `m` inside both for-loops in `smtRw` does not need to be mutable

Oh thanks. This is nicer.

**Comment by mhk119** [comment]
> Also, in `smtRw` you could iterate through arrays instead of indices to avoid unsafe operations, like:
> 
> ```lean
>   for innerArr in arr do
>     let m := innerArr.size
>     if m > 1 then
>       for j in [: m-1] do
>         let r ← mv'.rewrite (← mv'.getType) (mkAppN assoc #[innerArr[m-j-2]!]) true
>         mv' ← mv'.replaceTargetEq r.eNew r.eqProof
> ```
> 
> You could also iterate through the `Expr`s in `innerArr.reverse` instead of the indices. It looks like you are always skipping the last element of each array? if `j = 0` then `m - j - 2 = m - 2`, whereas the last element is at position `m - 1`

Yes. I am skipping the last element deliberately. This is also why I iterated over indices instead,

File: Smt/Reconstruction/Rewrites/Tactic.lean [hunk 1]
```diff
@@ -31,9 +31,16 @@ theorem bool_and_dup : (xs ∧ b ∧ ys ∧ b ∧ zs) = (xs ∧ b ∧ ys ∧ zs)
...
+def opsAssocNull : Name → Array Expr
+|  ``or => #[Expr.const ``or_assoc_eq [], Expr.const ``or_false []]
+| ``and => #[Expr.const ``and_assoc_eq [], Expr.const ``and_true []]
+| _ => #[]
+
+def smtRw (mv : MVarId) (op : Name) (rule : Expr) (arr : Array (Array Expr)) : MetaM Unit := do
   let n := arr.size
   let mut mv' := mv
+  let assoc := (opsAssocNull op)[0]!
+  let null := (opsAssocNull op)[1]!
```
[the remark this answers is not in the document]
Comment (mhk119) [reply Smt/Reconstruction/Rewrites/Tactic.lean:43]: Yes, I plan to add on more operators :).

**Pull request merged**